### PR TITLE
Improve dictionary operation

### PR DESCRIPTION
**Pull request by okuta**
Following code is inefficient in Cython.

``` python
if key in dict:
    return dict[key]
```

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Do you have some simple cupy snippets for profiling before/after? I see that a call to `__Pyx_PyDict_GetItem` is reduced but in `_kernel.pyx` the return statements became a bit more involved (since the object is now reassigned I think).

**Comment by okuta** [comment]
In Python2, string compare is very slow because Python2 convert raw string to UTF-8.

**Comment by hvy** [comment]
jenkins, test this please.

File: cupy/util.pyx [hunk 1]
```diff
@@ -41,10 +41,12 @@ def memoize(bint for_each_device=False):
             cdef dict m = memo
             if for_each_device:
                 id = device.get_device_id()
-            arg_key = (id, args, frozenset(kwargs.items()))
-            if arg_key in m:
-                result = m[arg_key]
+            if len(kwargs):
+                arg_key = (id, args)
             else:
+                arg_key = (id, args, frozenset(kwargs.items()))
```
Comment (hvy) [inline cupy/util.pyx]: Is the condition based on `kwargs` reversed, when assigning `arg_key`?

Comment (okuta) [reply cupy/util.pyx]: Oh!

**Comment by okuta** [comment]
I fixed.

File: cupy/util.pyx (hunk 1 above)
Comment (hvy) [reply cupy/util.pyx]: Thanks!

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Is this PR to be backported?

**Comment by hvy** [comment]
Jenkins timed out with 124. Rerunning tests.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Jenkins failure seems unrelated, finally... Thank you, LGTM.

**Pull request merged**